src/Swift.Runtime/src/Swift/SwiftString.cs

```diff
@@ -0,0 +1,164 @@
+
+    static SwiftString()
+    {
+        _protocolConformanceSymbols = new Dictionary<Type, string> { };
+    }
+
+    public static nuint PayloadSize => _payloadSize;
+
+    public Data Payload => _payload;
+
+    static TypeMetadata ISwiftObject.GetTypeMetadata()
+    {
+        return TypeMetadata.Cache.GetOrAdd(typeof(SwiftString), _ => PInvoke_getMetadata());
+    }
+
+    static ISwiftObject ISwiftObject.NewFromPayload(SwiftHandle handle)
+    {
+        return new SwiftString(handle);
+    }
+
+    IntPtr ISwiftObject.MarshalToSwift(IntPtr swiftDest)
+    {
+        var metadata = SwiftObjectHelper<SwiftString>.GetTypeMetadata();
+        unsafe
+        {
+            fixed (void* _payloadPtr = &_payload)
+            {
+                metadata.ValueWitnessTable->InitializeWithCopy((void*)swiftDest, (void*)_payloadPtr, metadata);
+            }
+        }
+        return swiftDest;
+    }
+
+    /// <summary>
+    /// Gets the protocol conformance descriptor for the given type.
+    /// </summary>
+    /// <typeparam name="TProtocol"></typeparam>
+    /// <returns></returns>
+    static ProtocolConformanceDescriptor ISwiftObject.GetProtocolConformanceDescriptor<TProtocol>()
+        where TProtocol : class
+    {
+        if (!_protocolConformanceSymbols.TryGetValue(typeof(TProtocol), out var symbolName))
+        {
+            throw new SwiftRuntimeException($"Attempted to retrieve protocol conformance descriptor for type SwiftString and protocol {typeof(TProtocol).Name}, but no conformance was found.");
+        }
+        return ProtocolConformanceDescriptor.LoadFromSymbol("/usr/lib/swift/libswiftCore.dylib", symbolName);
+    }
+
+    /// <summary>
+    /// Constructs a new SwiftString from the given handle.
+    /// </summary>
+    unsafe SwiftString(SwiftHandle handle)
+    {
+        _payload = *(Data*)handle;
+    }
+
+    /// <summary>
+    /// Constructs a new SwiftString from the C# string.
+    /// </summary>
+    public SwiftString(string str)
+    {
+        byte[] utf8Bytes = Encoding.UTF8.GetBytes(str);
+        unsafe
+        {
+            fixed (byte* utf8BytesPtr = utf8Bytes)
+            {
+                _payload = PInvoke_Create(utf8BytesPtr, utf8Bytes.Length, 1);
+            }
+        }
+    }
+
+    /// <summary>
+    /// Gets the length of string.
+    /// </summary>
+    public int Length => (int)PInvoke_GetLength(_payload);
+
+    /// <summary>
+    /// Converts the SwiftString to a C# string.
+    /// </summary>
+    public override string ToString()
+    {
+        unsafe
+        {
+            var elementType = TypeMetadata.GetTypeMetadataOrThrow<byte>();
+            var resultType = TypeMetadata.GetTypeMetadataOrThrow<long>();
+
+            var len = Length;
+            if (len <= 0)
+                return string.Empty;
+
+            string? str = null;
+
+            var arr = PInvoke_GetUtf8ContiguousArray(_payload);
+            PInvoke_WithUnsafeBytes(bytes =>
```

Comment:
How much do you care about performance? 

This will allocate a marshalled delegate on every call that is quite expensive operation. It would be a lot more efficient to use function pointers and use the `context` argument to pass the local context.

Reply:
Yes, this is a good point. We should implement the correct layout and memory management for the closure context. Also, we should be able to reduce the number of calls using a thin wrapper.

Created a tracking issue: https://github.com/dotnet/runtimelab/issues/2990

Reply:
Doing this properly using function pointers is like 10-line edit on the new code added in this PR...

Reply:
Ok, let's fix that in this PR.